presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneDistinctAggregation.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.ExceptNode;
+import io.prestosql.sql.planner.plan.IntersectNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.PlanVisitor;
+import io.prestosql.sql.planner.plan.UnionNode;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.sql.planner.plan.ChildReplacer.replaceChildren;
+import static io.prestosql.sql.planner.plan.Patterns.aggregation;
+
+public class PruneDistinctAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(PruneDistinctAggregation::isDistinctOperator);
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode node, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        DistinctAggregationRewriter rewriter = new DistinctAggregationRewriter(lookup);
+
+        List<PlanNode> newSources = node.getSources().stream()
+                .flatMap(lookup::resolveGroup)
+                .map(source -> source.accept(rewriter, true))
+                .collect(Collectors.toList());
+
+        if (rewriter.isRewritten()) {
+            return Result.ofPlanNode(replaceChildren(node, newSources));
+        }
+        else {
+            return Result.empty();
+        }
+    }
+
+    private static boolean isDistinctOperator(AggregationNode node)
+    {
+        return node.getAggregations().isEmpty();
+    }
+
+    private static class DistinctAggregationRewriter
```

Comment:
Instead of using a visitor, this should be done pairwise by a rule that matches the expected pattern. I.e., it if matches `Aggregation -> SetOperation -> {Aggregation -> X, ... }`, it should rewrite to `Aggregation -> SetOperation -> {X, ...}`.

Reply:
I see. But it needs to match any number of the intermediate set operations as [described before](https://github.com/prestosql/presto/pull/2619#issuecomment-583388309).

It should rewrite the following patterns.
```
Aggregation -> SetOperation -> {Aggregation -> X, ...
Aggregation -> SetOperation -> {SetOperation -> {Aggregation -> X, ...
Aggregation -> SetOperation -> {SetOperation -> {SetOperation -> {Aggregation -> X, ...
...
```

Initially, I wrote the `DistinctAggregationRewriter` as pattern match rule. But it did not work as expected due to the previous reason. Is it possible to match the pattern with arbitrary number of intermediate set operations?

Reply:
That's right. Currently it's not possible to match that pattern with rules. If the intermediate set operations are of the same type, the MergeXXX rules should take care of them, but if they are not, there's no way to jump an arbitrary number of levels. This will require support for traits and properties in the iterative optimizer. Let's leave it as you did for now.